api/agent/drivers/docker/imageCache.go

```diff
@@ -0,0 +1,210 @@
+/*
+ImageCache holds all the logic for calculating what docker images can be removed from the running agent.
+The last used time and the number of uses are both taken into account to calculate a score (timeSinceLastUse/uses)
+The higher the score the more evicitable the image is.
+
+ImageCache also provides a method to "lock" an image, insuring it is never deleted. To do so a Lock is called with
+the image ID to lock, as well as a token. The token is then added to a set of tokens attached to that entry.
+The set is a map of *interface -> *interface where both values are the same.
+*/
+
+package docker
+
+import (
+	"errors"
+	"sort"
+	"sync"
+	"time"
+
+	d "github.com/fsouza/go-dockerclient"
+	"github.com/sirupsen/logrus"
+)
+
+// Cache is a list of Entries. Stored as type EntryByAge so that sort works.
+// The mutex is present to control access access the list of images and their locks
+// The Cache may be used in concurrent contexts
+type Cache struct {
```

Comment:
probably want to either make all of these objects and methods private or make a package off of the drivers package with this stuff like `drivers/lrucache` where we could theoretically lru cache any driver's images (i don't want to think about pluggable driver image remove methods really at this hour though, up to you)